colossalai/shardformer/policies/bert.py

```diff
@@ -275,11 +287,28 @@ def get_held_layers(self) -> List[Module]:
         stage_manager = self.pipeline_stage_manager
 
         held_layers = []
-        layers_per_stage = self.distribute_layers(len(module.encoder.layer), stage_manager.num_stages)
+        num_model_chunks = stage_manager.num_model_chunks
+        layers_per_stage = self.distribute_layers(
+            len(module.encoder.layer), stage_manager.num_stages * num_model_chunks
+        )
         if stage_manager.is_first_stage():
             held_layers.append(module.embeddings)
-        start_idx, end_idx = self.get_stage_index(layers_per_stage, stage_manager.stage)
-        held_layers.extend(module.encoder.layer[start_idx:end_idx])
+        if num_model_chunks > 1:
+            stage_index = Policy.get_stage_index(
+                layers_per_stage, stage_manager.stage, stage_manager.num_stages, num_model_chunks
+            )
+        else:
+            stage_index = Policy.get_stage_index(layers_per_stage, stage_manager.stage)
+
+        # interleaved stage index for one device comes in pairs, e.g.[[0,3],[6,9]]
```

Comment:
this illustrates the main idea for interleaved pipeline parallel. take bert for example, the original non-interlaved bert layer distribution among the two pipeline parallel devices are: device0=[0,6], device1=[6,12]; for interleaved, we now have: device0=[[0,3],[6,9]], deivce1=[[3,6],[9,12]]